doc/rtd/howto/return_codes.rst

```diff
@@ -0,0 +1,172 @@
+.. _return_codes:
+
+Why did `cloud-init status` start returning exit code 2?
+========================================================
+
+Cloud-init introduced a new error code in 23.4.
+This page describes the purpose of this change and gives
+some context for why this change was made.
+
+.. _return_codes_history:
+
+History
+-------
+
+In the early days of cloud-init, the purpose of cloud-init
+was simple: provision ssh keys and a user and exit. Cloud-init
+only supported a couple of clouds, and things were simple.
+Since cloud-init provides access to cloud instances, the
+paradigm for handling errors was "log errors, but proceed".
+Exiting on failure conditions doesn't make sense when that
+may prevent one from accessing the system to debug it.
+
+Since cloud-init's behavior is heavily tied to specific cloud
+platforms, reproducing cloud-init bugs without exactly
+reproducing a specific cloud environment is often impossible,
+and often requires guesswork. To make debugging cloud-init
+possible without reproducing exactly, cloud-init logs are
+quite verbose.
+
+.. _return_codes_pain_points:
+
+Pain points
+-----------
+
+1) Invalid configurations were historically "ignored".
+
+2) Log verbosity is unfriendly to end users that may not know
+  what to look for. Verbose logs means users often ignore real
+  errors.
+
+3) Cloud-init's reported status was only capable of telling the user
+  whether cloud-init crashed. If a user's configuration was invalid,
+  or if the operating system or cloud environment experienced some
+  error that prevented cloud-init from configuring the instance, or
+  if cloud-init internally experienced an error - all of these
+  previously reported a status of "done".
+
+.. _return_codes_improvements
+
+Efforts to improve cloud-init
+-----------------------------
+
+Several changes have been introduced to cloud-init to address the pain
+points described above.
+
+Jsonschema
+^^^^^^^^^^
+
+Cloud-init has defined a jsonschema which fully documents the user-data
+cloud-config. This jsonschema may be used in several different ways:
+
+Text editor integration
+"""""""""""""""""""""""
+
+Thanks to `yamlls<yaml-language-server>`, cloud-init's jsonschema may be
+used for yaml syntax checking, warnings when invalid keys are used, and
+even autocompletion. Several different text editors are capable of this.
+See this `blog post on configuring this for neovim`_, or 
+
+# TODO someone who knows about vscode add some details
+
+Cloud-init schema subcommand
+""""""""""""""""""""""""""""
+
+The cloud-init package includes a cloud-init subcommand, `cloud-init schema`
```

Comment:
Should link to https://cloudinit.readthedocs.io/en/latest/howto/debug_user_data.html